src/Sylius/Bundle/SettingsBundle/Doctrine/EventListener/ParameterTransformerListener.php

```diff
@@ -0,0 +1,132 @@
+
+    /**
+     * @var array
+     */
+    protected $parametersMap = [];
+
+    public function __construct(ContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
+    /**
+     * @param LifecycleEventArgs $event
+     */
+    public function postLoad(LifecycleEventArgs $event)
+    {
+        $settings = $event->getObject();
+
+        if ($settings instanceof SettingsInterface) {
+            $this->reverseTransform($settings);
+        }
+    }
+
+    /**
+     * @param OnFlushEventArgs $event
+     */
+    public function onFlush(OnFlushEventArgs $event)
+    {
+        $em = $event->getEntityManager();
+        $uow = $em->getUnitOfWork();
+
+        foreach ($uow->getScheduledEntityInsertions() as $entity) {
+            if ($entity instanceof SettingsInterface) {
+                $this->transform($entity);
+            }
+        }
+
+        foreach ($uow->getScheduledEntityUpdates() as $entity) {
+            if ($entity instanceof SettingsInterface) {
+                $this->transform($entity);
+            }
+        }
+    }
+
+    public function postFlush()
+    {
+        // revert settings parameters to what they were before flushing
+        foreach ($this->parametersMap as $map) {
+            $map['entity']->setParameters($map['parameters']);
+        }
+
+        // reset parameters map
+        $this->parametersMap = [];
+    }
+
+    /**
+     * @param SettingsInterface $settings
+     */
+    protected function transform(SettingsInterface $settings)
+    {
+        // store old parameters, so we can revert to it after flush
+        $this->parametersMap[] = [
+            'entity' => $settings,
+            'parameters' => $settings->getParameters(),
+        ];
+
+        $transformers = $this->getTransformers($settings);
+        foreach ($settings->getParameters() as $name => $value) {
+            if (isset($transformers[$name])) {
+                $settings->set($name, $transformers[$name]->transform($value));
+            }
+        }
+    }
+
+    /**
+     * @param SettingsInterface $settings
+     */
+    protected function reverseTransform(SettingsInterface $settings)
+    {
+        $transformers = $this->getTransformers($settings);
+        foreach ($settings->getParameters() as $name => $value) {
+            if (isset($transformers[$name])) {
+                $settings->set($name, $transformers[$name]->reverseTransform($value));
+            }
+        }
+    }
+
+    /**
+     * @param SettingsInterface $settings
+     *
+     * @return ParameterTransformerInterface[]
+     */
+    protected function getTransformers(SettingsInterface $settings)
+    {
+        $registry = $this->container->get('sylius.settings.schema_registry');
+        $schema = $registry->get($settings->getSchemaAlias());
+
+        $settingsBuilder = new SettingsBuilder();
+        $schema->buildSettings($settingsBuilder);
+
+        return $settingsBuilder->getTransformers();
+    }
+}
```

Comment:
Well theme settings is a good use case, also the only one I can think of. With themes you don't really want to roll out your own separate settings management system, so re-using SettingsBundle is really good here. But if we want to use SettingsBundle with theme settings you want to be able to transform setting values. This is because themes can have settings that sets a specific `Taxon` for example, or a theme setting where you can link a `Page` to it. With data transformers on the settings bundle level it is very easy to implement all that for themes. Combine this with the resolver based schema loading in this PR and I think you can build settings for the craziest things. 

I am very biased to themes, but that is because I wanted to re-use the settings bundle for my own theming system. But I think it a good use case for Sylius too.

About the performance, I don't know if it's much overhead, I haven't profiled it yet. I think it is negligible, but I may be wrong here.